### PR TITLE
enhance copy to reuse buffer

### PR DESCRIPTION
**Pull request by zkidkid**
Problem

Enhance StreamIO::copy function by re-use tmp buffer.

Solution

Create temp buffer before calling recursive function.

### CODE REVIEW

File: util-core/src/main/scala/com/twitter/io/StreamIO.scala [hunk 1]
```diff
@@ -6,28 +6,36 @@ import scala.annotation.tailrec
...
+    * buffer size (default = 1KB).
+    */
   final def copy(
-    inputStream:  InputStream,
-    outputStream: OutputStream,
-    bufferSize:   Int = 1024
-  ) {
-    val buf = new Array[Byte](bufferSize)
-    inputStream.read(buf, 0, buf.length) match {
+                  inputStream: InputStream,
```
Comment (mosesn) [inline util-core/src/main/scala/com/twitter/io/StreamIO.scala]: only two spaces please

File: util-core/src/main/scala/com/twitter/io/StreamIO.scala [hunk 2]
```diff
@@ -6,28 +6,36 @@ import scala.annotation.tailrec
...
+                  outputStream: OutputStream,
+                  bufferSize: Int = 1024
+                ) {
+    val buffer = new Array[Byte](bufferSize)
+    copy(inputStream, outputStream, buffer)
+  }
+
+  @tailrec
+  final private def copy(
+                  inputStream: InputStream,
```
Comment (mosesn) [inline util-core/src/main/scala/com/twitter/io/StreamIO.scala]: only two spaces please

File: util-core/src/main/scala/com/twitter/io/StreamIO.scala [hunk 3]
```diff
@@ -6,28 +6,36 @@ import scala.annotation.tailrec
 
 object StreamIO {
   /**
-   * Copy an InputStream to an OutputStream in chunks of the given
-   * buffer size (default = 1KB).
-   */
-  @tailrec
+    * Copy an InputStream to an OutputStream in chunks of the given
```
Comment (mosesn) [inline util-core/src/main/scala/com/twitter/io/StreamIO.scala]: keep stars aligned to the left star (-1 col)

File: util-core/src/main/scala/com/twitter/io/StreamIO.scala [hunk 4]
```diff
@@ -6,28 +6,36 @@ import scala.annotation.tailrec
...
+        outputStream.write(buffer, 0, n)
+        copy(inputStream, outputStream, buffer)
     }
   }
 
   /**
-   * Buffer (fully) the given input stream by creating & copying it to
-   * a ByteArrayOutputStream.
-   */
+    * Buffer (fully) the given input stream by creating & copying it to
```
Comment (mosesn) [inline util-core/src/main/scala/com/twitter/io/StreamIO.scala]: keep stars aligned to the left star (-1 col)

**Comment by kevinoliver** [comment]
nice find. looks good to me after applying the formatting changes from @mosesn

File: util-core/src/main/scala/com/twitter/io/StreamIO.scala [hunk 5]
```diff
@@ -6,39 +6,39 @@ import scala.annotation.tailrec
...
+   */
   final def copy(
-                  inputStream: InputStream,
-                  outputStream: OutputStream,
-                  bufferSize: Int = 1024
-                ) {
+    inputStream: InputStream,
+    outputStream: OutputStream,
+    bufferSize: Int = 1024
+  ) {
```
Comment (taylorleese) [inline util-core/src/main/scala/com/twitter/io/StreamIO.scala]: Please add return a type ```: Unit = ```.

File: util-core/src/main/scala/com/twitter/io/StreamIO.scala [hunk 6]
```diff
@@ -6,39 +6,39 @@ import scala.annotation.tailrec
...
   @tailrec
   final private def copy(
-                  inputStream: InputStream,
-                  outputStream: OutputStream,
-                  buffer: Array[Byte]
-                ) {
+    inputStream: InputStream,
+    outputStream: OutputStream,
+    buffer: Array[Byte]
+  ) {
```
Comment (taylorleese) [inline util-core/src/main/scala/com/twitter/io/StreamIO.scala]: Please add return a type ```: Unit = ```.

**Comment by mosesn** [comment]
LGTM after @taylorleese's nits.

**Comment by kevinoliver** [comment]
Thanks @zkidkid I'll get this merged in.

**Comment by kevinoliver** [comment]
Ok this got merged internally and should show up on the `develop` branch in the next push. Thanks again for the patch, @zkidkid.

**Comment by kevinoliver** [comment]
Merged in b06cef2c02a78268f13e46d001b40bce61d7f042

**Pull request closed without merging**